File: src/core/Stocks/Services/GapAnalysis.cs
#nullable enable
using System;
using System.Collections.Generic;
using core.Shared.Adapters.Stocks;

namespace core.Stocks.Services
{
    public class GapAnalysis
    {
        public static List<Gap> Generate(Span<PriceBar> prices, int numberOfBarsToAnalyze)
        {
            var start = prices.Length > numberOfBarsToAnalyze
                ? prices.Length - numberOfBarsToAnalyze
                : 0;
            return Generate(prices.Slice(start));
        }

        public static List<Gap> Generate(Span<PriceBar> prices)
        {
            var gaps = new List<Gap>();

            for (var i = 1; i < prices.Length; i++)
            {
                var yesterday = prices[i - 1];
                var currentBar = prices[i];

                var gapSizePct = 0m;

                if (currentBar.Low > yesterday.High)
                {
                    gapSizePct = Math.Round( (currentBar.Low - yesterday.High)/yesterday.High * 100, 2);
                }
                else if (currentBar.High < yesterday.Low)
                {
                    gapSizePct = -1 * Math.Round( (yesterday.Low - currentBar.High)/yesterday.Low * 100, 2);
                }

                if (gapSizePct != 0)
                {
                    var type = gapSizePct switch {
                        > 0 => GapType.Up,
                        < 0 => GapType.Down,
                        _ => throw new Exception("Invalid gap type")
                    };
                    var percentChange = Math.Round( (currentBar.Close - yesterday.Close)/yesterday.Close * 100, 2);

                    Func<PriceBar, bool> closingCondition = gapSizePct switch {
                        > 0 => bar => bar.Close <= yesterday.Close,
                        < 0 => bar => bar.Close >= yesterday.Close,
                        _ => throw new Exception("Invalid gap type")
                    };

                    var closedQuickly = ClosingConditionMet(prices, i + 1, 10, closingCondition);
                    var open = !ClosingConditionMet(prices, i + 1, prices.Length - i, closingCondition);

                    var gap = new Gap(
                        type: type,
                        gapSizePct: gapSizePct,
                        percentChange: percentChange,
                        bar: currentBar,
                        closedQuickly: closedQuickly,
                        open: open
                    );
                    gaps.Add(gap);
                }
            }

            return gaps;
        }

        private static bool ClosingConditionMet(
            Span<PriceBar> prices,
            int start, int length,
            Func<PriceBar, bool> closingCondition)
        {
            for(var i = start; i < start + length && i < prices.Length; i++)
            {
                if (closingCondition(prices[i]))
                {
                    return true;
                }
            }
            return false;
        }
    }

    public record struct Gap(
        GapType type,
        decimal gapSizePct,
        decimal percentChange,
        PriceBar bar,
        bool closedQuickly,
        bool open);

    public enum GapType { Up, Down }
}
#nullable restore

File: src/core.fs/Services/GapAnalysis.fs
module core.fs.Services.GapAnalysis

open core.fs.Adapters.Stocks
open core.fs.Services.Analysis

    
type GapType =
    | Up
    | Down
    
    static member FromString(value:string) =
        match value with
        | nameof Up -> Up
        | nameof Down -> Down
        | _ -> failwith $"Invalid gap type: {value}"
        
    override this.ToString() =
        match this with
        | Up -> nameof Up
        | Down -> nameof Down
        
type Gap =
    {
        Type: GapType
        GapSizePct: decimal
        PercentChange: decimal
        Bar: PriceBar
        ClosedQuickly: bool
        Open: bool
        RelativeVolume: decimal option
        ClosingRange: decimal
    }
        
let private closingConditionMet (prices: PriceBar array) (start: int) (length: int) (closingCondition: PriceBar -> bool) =
    
    let rec loop (i: int) =
        if i < start + length && i < prices.Length then
            if closingCondition prices[i] then
                true
            else
                loop (i + 1)
        else
            false
            
    loop start
        
let private generateInternal (prices: PriceBars) (volumeStats: core.fs.Services.Analysis.DistributionStatistics) =
    
    prices.Bars
    |> Array.pairwise
    |> Array.filter (fun (previous, current) -> current.HasGap(previous))
    |> Array.indexed
    |> Array.map ( fun (index, (yesterday, current)) ->
        let gapSizePct =
            let referencePrice = min current.Open current.Close
            (referencePrice - yesterday.Close) / yesterday.Close
        
        let type' =
            match gapSizePct with
            | x when x > 0m -> GapType.Up
            | x when x < 0m -> GapType.Down
            | _ -> failwith $"Invalid gap type: {current} vs {yesterday}"
            
        let percentChange = (current.Close - yesterday.Close) / yesterday.Close
        
        let closingCondition =
            match gapSizePct with
            | x when x > 0m -> fun (bar:PriceBar) -> bar.Close <= yesterday.Close
            | x when x < 0m -> fun bar -> bar.Close >= yesterday.Close
            | _ -> failwith "Invalid gap type"
            
        let i = index * 2 // each indexed pair has two bars
        let closedQuickly = closingConditionMet prices.Bars i 10 closingCondition
        let open' = not (closingConditionMet prices.Bars i (prices.Length - i) closingCondition)
        let relativeVolume =
            match volumeStats.count < Constants.NumberOfDaysForRecentAnalysis with
            | false -> System.Math.Round ((current.Volume |> decimal) / volumeStats.mean, 2) |> Some
            | true -> None
            
        let closingRange = current.ClosingRange()
        
        {
            Type = type'
            GapSizePct = gapSizePct
            PercentChange = percentChange
            Bar = current
            ClosedQuickly = closedQuickly
            Open = open'
            RelativeVolume = relativeVolume
            ClosingRange = closingRange
        }
    )
    
let detectGaps (numberOfBarsToAnalyze: int) (prices: PriceBars) =
    
    let barsForAnalysis = numberOfBarsToAnalyze |> prices.LatestOrAll
    let barsForVolume = Constants.NumberOfDaysForRecentAnalysis |> prices.LatestOrAll 
    let volumeStats = barsForVolume.Volumes() |> DistributionStatistics.calculate
    
    generateInternal barsForAnalysis volumeStats
